File: ios/bitchat/bitchat/V2V/Services/V2VEmergencyService.swift
import Foundation
import CoreLocation
import BitFoundation

public protocol V2VEmergencyDelegate: AnyObject {
    func onEmergencyAlertReceived(_ alert: ReceivedAlert)
    func onEmergencyBroadcastStarted(vehicleType: VehicleType)
    func onEmergencyBroadcastStopped()
}

public protocol V2VMeshBroadcaster: AnyObject {

    var myPeerId: String { get }

    func broadcastEmergencyAlert(payload: Data, ttl: UInt8)
}

public final class V2VEmergencyService: NSObject, CLLocationManagerDelegate {

    private static let broadcastIntervalMs: Int64 = 1_000
    private static let locationUpdateMinMeters: CLLocationDistance = 1
    private static let emergencyTtl: UInt8 = 7

    public weak var delegate: V2VEmergencyDelegate?

    private let meshBroadcaster: V2VMeshBroadcaster
    private let firebaseTransport: FirebaseTransport
    private let logger = TransportLogRepository.shared
    private let deduplicator = AlertDeduplicationService()

    private let locationManager = CLLocationManager()
    public private(set) var currentLocation: CLLocation?
    public private(set) var currentSpeed: Float = 0
    public private(set) var currentHeading: Float = 0

    public var onLocationUpdate: ((CLLocation) -> Void)?

    public private(set) var currentVehicleType: VehicleType = .ambulance
    public private(set) var isEmergencyActive = false
    public private(set) var receivedAlerts: [ReceivedAlert] = []

    private var broadcastTimer: DispatchSourceTimer?
    private let broadcastQueue = DispatchQueue(label: "v2v.broadcast")

    public init(meshBroadcaster: V2VMeshBroadcaster, firebaseTransport: FirebaseTransport = .shared) {
        self.meshBroadcaster = meshBroadcaster
        self.firebaseTransport = firebaseTransport
        super.init()
        setupLocation()
    }

    deinit {
        shutdown()
    }

    public func shutdown() {
        stopEmergencyBroadcast()
        locationManager.stopUpdatingLocation()
        locationManager.stopUpdatingHeading()
    }

    private func setupLocation() {
        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyBest
        locationManager.distanceFilter = Self.locationUpdateMinMeters
        locationManager.activityType = .automotiveNavigation

        locationManager.startUpdatingLocation()
        if CLLocationManager.headingAvailable() {
            locationManager.startUpdatingHeading()
        }
    }

    public func setVehicleType(_ type: VehicleType) {
        currentVehicleType = type
    }

    public func startEmergencyBroadcast(vehicleType: VehicleType? = nil) {
        guard !isEmergencyActive else { return }
        if let vt = vehicleType { currentVehicleType = vt }
        isEmergencyActive = true

        let timer = DispatchSource.makeTimerSource(queue: broadcastQueue)
        timer.schedule(deadline: .now(), repeating: .milliseconds(Int(Self.broadcastIntervalMs)))
        timer.setEventHandler { [weak self] in
            guard let self = self, self.isEmergencyActive else { return }
            guard let location = self.currentLocation else { return }
            self.broadcastEmergencyAlert(from: location)
        }
        timer.resume()
        broadcastTimer = timer

        delegate?.onEmergencyBroadcastStarted(vehicleType: currentVehicleType)
    }

    public func stopEmergencyBroadcast() {
        guard isEmergencyActive else { return }
        isEmergencyActive = false
        broadcastTimer?.cancel()
        broadcastTimer = nil
        delegate?.onEmergencyBroadcastStopped()
    }

    public func toggleEmergencyBroadcast() {
        if isEmergencyActive { stopEmergencyBroadcast() } else { startEmergencyBroadcast() }
    }

    private func broadcastEmergencyAlert(from location: CLLocation) {
        let alertType: AlertType
        switch Float(location.speed) {
        case ..<1:  alertType = .stationary
        case ..<5:  alertType = .passing
        default:    alertType = .approaching
        }
        let alert = V2VEmergencyAlert(
            vehicleType: currentVehicleType,
            alertType: alertType,
            latitude: location.coordinate.latitude,
            longitude: location.coordinate.longitude,
            speed: Float(max(location.speed, 0)),
            heading: Float(max(location.course, 0)),
            senderPeerId: meshBroadcaster.myPeerId
        )

        let bleStart = Date()
        let payload = alert.toPayload()
        meshBroadcaster.broadcastEmergencyAlert(payload: payload, ttl: Self.emergencyTtl)
        let bleLatency = Int64(Date().timeIntervalSince(bleStart) * 1000)
        logger.logSend(
            transport: .ble,
            messageId: alert.messageId,
            latencyMs: bleLatency,
            payloadBytes: payload.count,
            details: "peer=\(meshBroadcaster.myPeerId.prefix(8))"
        )

        Task {
            let fbStart = Date()
            do {
                guard firebaseTransport.isAvailable else { return }
                let cloudAlert = FirebaseEmergencyAlert(from: alert)
                try await firebaseTransport.sendEmergencyAlert(cloudAlert)
                let fbLatency = Int64(Date().timeIntervalSince(fbStart) * 1000)
                self.logger.logSend(
                    transport: .firebase,
                    messageId: alert.messageId,
                    latencyMs: fbLatency,
                    payloadBytes: cloudAlert.toBytes().count,
                    details: "channel=emergency"
                )
            } catch {
                self.logger.logFailure(
                    transport: .firebase,
                    direction: .send,
                    messageId: alert.messageId,
                    details: "Firebase send error: \(error.localizedDescription)"
                )
            }
        }
    }

    @discardableResult
    public func processIncomingPayload(
        _ payload: Data,
        fromPeerId: String,
        sentAtMs: Int64? = nil,
        transport: TransportType = .ble
    ) -> V2VEmergencyAlert? {
        let receiveTimestamp = Int64(Date().timeIntervalSince1970 * 1000)
        guard let alert = V2VEmergencyAlert.fromPayload(payload) else {
            logger.logFailure(
                transport: transport,
                direction: .receive,
                messageId: "unknown",
                details: "Decode failed from \(fromPeerId)"
            )
            return nil
        }
        guard deduplicator.checkAndMark(alert.messageId) else { return nil }

        let latency = receiveTimestamp - (sentAtMs ?? alert.timestamp)
        logger.logReceive(
            transport: transport,
            messageId: alert.messageId,
            latencyMs: latency,
            payloadBytes: payload.count,
            details: "from=\(fromPeerId.prefix(8))"
        )

        let distance: Float
        let direction: String
        if let me = currentLocation {
            let alertLoc = CLLocation(latitude: alert.latitude, longitude: alert.longitude)
            distance = Float(me.distance(from: alertLoc))
            direction = relativeDirection(myLocation: me, alert: alert)
        } else {
            distance = Float.greatestFiniteMagnitude
            direction = ""
        }

        let received = ReceivedAlert(
            alert: alert,
            distanceMeters: distance,
            relativeDirection: direction
        )
        addReceivedAlert(received)
        delegate?.onEmergencyAlertReceived(received)
        return alert
    }

    public func cleanupExpiredAlerts() {
        receivedAlerts.removeAll { !$0.isValid }
    }

    public func getDeduplicationService() -> AlertDeduplicationService { deduplicator }

    private func addReceivedAlert(_ alert: ReceivedAlert) {
        receivedAlerts.removeAll { !$0.isValid }

        receivedAlerts.removeAll { $0.alert.senderPeerId == alert.alert.senderPeerId }
        receivedAlerts.append(alert)
        receivedAlerts.sort { $0.distanceMeters < $1.distanceMeters }
    }

    private func relativeDirection(myLocation: CLLocation, alert: V2VEmergencyAlert) -> String {
        let bearing = computeBearing(
            lat1: myLocation.coordinate.latitude,
            lon1: myLocation.coordinate.longitude,
            lat2: alert.latitude,
            lon2: alert.longitude
        )
        let myHeading = Double(currentHeading)
        var relative = bearing - myHeading
        if relative < 0 { relative += 360 }
        if relative > 180 { relative -= 360 }
        switch relative {
        case -45...45:    return "ahead"
        case 45...135:    return "right"
        case -135 ... -45: return "left"
        default:          return "behind"
        }
    }

    private func computeBearing(lat1: Double, lon1: Double, lat2: Double, lon2: Double) -> Double {
        let lat1Rad = lat1 * .pi / 180
        let lat2Rad = lat2 * .pi / 180
        let dLon = (lon2 - lon1) * .pi / 180
        let x = sin(dLon) * cos(lat2Rad)
        let y = cos(lat1Rad) * sin(lat2Rad) - sin(lat1Rad) * cos(lat2Rad) * cos(dLon)
        var bearing = atan2(x, y) * 180 / .pi
        if bearing < 0 { bearing += 360 }
        return bearing
    }

    public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else { return }
        currentLocation = location
        currentSpeed = Float(max(location.speed, 0))

        if location.course >= 0 {
            currentHeading = Float(location.course)
        }
        onLocationUpdate?(location)
    }

    public func locationManager(_ manager: CLLocationManager, didUpdateHeading newHeading: CLHeading) {
        if newHeading.trueHeading >= 0 {
            currentHeading = Float(newHeading.trueHeading)
        }
    }

    public func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {

    }
}

File: app/src/main/java/com/cybersiren/android/v2v/service/V2VEmergencyService.kt
package com.cybersiren.android.v2v.service

import android.content.Context
import android.location.Location
import android.util.Log
import com.cybersiren.android.geohash.LocationProvider
import com.cybersiren.android.mesh.BluetoothMeshService
import com.cybersiren.android.protocol.BitchatPacket
import com.cybersiren.android.protocol.MessageType
import com.cybersiren.android.protocol.SpecialRecipients
import com.cybersiren.android.v2v.model.*
import com.cybersiren.android.v2v.model.TransportType
import com.cybersiren.android.v2v.model.TransportDirection
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import com.cybersiren.android.online.FirebaseTransport
import com.cybersiren.android.online.EmergencyAlert as FirebaseEmergencyAlert

class V2VEmergencyService(
    private val context: Context,
    private val meshService: BluetoothMeshService,
    private val locationProvider: LocationProvider,
    private val myPeerId: String
) {
    companion object {
        private const val TAG = "V2VEmergencyService"
        private const val BROADCAST_INTERVAL_MS = 1000L
        private const val LOCATION_UPDATE_INTERVAL_MS = 500L
        private const val LOCATION_MIN_DISTANCE_M = 1f
        private const val EMERGENCY_TTL: UByte = 7u
    }

    private val deduplicationService = AlertDeduplicationService()

    private val serviceScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private var broadcastJob: Job? = null
    private var locationUpdateJob: Job? = null

    private val _isEmergencyActive = MutableStateFlow(false)
    val isEmergencyActive: StateFlow<Boolean> = _isEmergencyActive.asStateFlow()

    private val _currentVehicleType = MutableStateFlow(VehicleType.AMBULANCE)
    val currentVehicleType: StateFlow<VehicleType> = _currentVehicleType.asStateFlow()

    private val _currentLocation = MutableStateFlow<Location?>(null)
    val currentLocation: StateFlow<Location?> = _currentLocation.asStateFlow()

    private val _currentSpeed = MutableStateFlow(0f)
    val currentSpeed: StateFlow<Float> = _currentSpeed.asStateFlow()

    private val _currentHeading = MutableStateFlow(0f)
    val currentHeading: StateFlow<Float> = _currentHeading.asStateFlow()

    private val _receivedAlerts = MutableStateFlow<List<ReceivedAlert>>(emptyList())
    val receivedAlerts: StateFlow<List<ReceivedAlert>> = _receivedAlerts.asStateFlow()

    var delegate: V2VEmergencyDelegate? = null

    private var locationCallback: ((Location) -> Unit)? = null

    init {
        startLocationUpdates()
    }

    private fun startLocationUpdates() {
        locationCallback = { location ->
            _currentLocation.value = location
            _currentSpeed.value = location.speed
            _currentHeading.value = location.bearing
        }

        locationCallback?.let { callback ->
            locationProvider.requestLocationUpdates(
                intervalMs = LOCATION_UPDATE_INTERVAL_MS,
                minDistanceMeters = LOCATION_MIN_DISTANCE_M,
                callback = callback
            )
        }

        Log.d(TAG, "Location updates started")
    }

    fun setVehicleType(vehicleType: VehicleType) {
        _currentVehicleType.value = vehicleType
        Log.d(TAG, "Vehicle type set to: ${vehicleType.displayName}")
    }

    fun startEmergencyBroadcast(vehicleType: VehicleType = _currentVehicleType.value) {
        if (_isEmergencyActive.value) {
            Log.w(TAG, "Emergency broadcast already active")
            return
        }

        _currentVehicleType.value = vehicleType
        _isEmergencyActive.value = true

        broadcastJob = serviceScope.launch {
            Log.i(TAG, "Starting emergency broadcast as ${vehicleType.displayName}")

            while (isActive && _isEmergencyActive.value) {
                val location = _currentLocation.value
                if (location != null) {
                    broadcastEmergencyAlert(location)
                } else {
                    Log.w(TAG, "No location available for broadcast")
                }
                delay(BROADCAST_INTERVAL_MS)
            }
        }

        delegate?.onEmergencyBroadcastStarted(vehicleType)
    }

    fun stopEmergencyBroadcast() {
        if (!_isEmergencyActive.value) {
            Log.w(TAG, "Emergency broadcast not active")
            return
        }

        _isEmergencyActive.value = false
        broadcastJob?.cancel()
        broadcastJob = null

        Log.i(TAG, "Emergency broadcast stopped")
        delegate?.onEmergencyBroadcastStopped()
    }

    fun toggleEmergencyBroadcast(): Boolean {
        if (_isEmergencyActive.value) {
            stopEmergencyBroadcast()
        } else {
            startEmergencyBroadcast()
        }
        return _isEmergencyActive.value
    }

    private fun broadcastEmergencyAlert(location: Location) {
        val alert = EmergencyAlert(
            vehicleType = _currentVehicleType.value,
            alertType = determineAlertType(location),
            latitude = location.latitude,
            longitude = location.longitude,
            speed = location.speed,
            heading = location.bearing,
            senderPeerId = myPeerId
        )

        val packet = createAlertPacket(alert)
        val bleStart = System.currentTimeMillis()
        try {
            meshService.broadcastPacket(packet)
            val bleLatency = System.currentTimeMillis() - bleStart
            TransportLogRepository.logSend(
                transport = TransportType.BLE,
                messageId = alert.messageId,
                latencyMs = bleLatency,
                payloadBytes = packet.payload.size,
                details = "peer=${myPeerId.take(8)}"
            )
            Log.d(TAG, "V2V: Broadcast via BLE Mesh (${bleLatency}ms)")
        } catch (e: Exception) {
            TransportLogRepository.logFailure(
                transport = TransportType.BLE,
                direction = TransportDirection.SEND,
                messageId = alert.messageId,
                details = "BLE send error: ${e.message}"
            )
            Log.w(TAG, "V2V: BLE send failed: ${e.message}")
        }

        serviceScope.launch {
            val fbStart = System.currentTimeMillis()
            try {
                val firebaseTransport = FirebaseTransport.getInstance(context)
                if (firebaseTransport.isAvailable) {

                    val firebaseAlert = FirebaseEmergencyAlert(
                        messageId = alert.messageId,
                        type = FirebaseEmergencyAlert.AlertType.EMERGENCY,
                        vehicleType = when (alert.vehicleType) {
                            VehicleType.AMBULANCE -> FirebaseEmergencyAlert.VehicleType.AMBULANCE
                            VehicleType.POLICE_CAR -> FirebaseEmergencyAlert.VehicleType.POLICE
                            VehicleType.FIRE_TRUCK -> FirebaseEmergencyAlert.VehicleType.FIRE_TRUCK
                            VehicleType.EMERGENCY -> FirebaseEmergencyAlert.VehicleType.OTHER
                        },
                        latitude = alert.latitude,
                        longitude = alert.longitude,
                        speedKmh = (alert.speed * 3.6).toInt(),
                        heading = alert.heading.toInt()
                    )

                    firebaseTransport.sendEmergencyAlert(firebaseAlert)
                    val fbLatency = System.currentTimeMillis() - fbStart
                    TransportLogRepository.logSend(
                        transport = TransportType.FIREBASE,
                        messageId = alert.messageId,
                        latencyMs = fbLatency,
                        payloadBytes = alert.toPayload().size,
                        details = "channel=emergency"
                    )
                    Log.i(TAG, "V2N: Broadcast via Firebase Cloud (${fbLatency}ms)")
                } else {
                    Log.d(TAG, "V2N: Firebase not available (offline mode)")
                }
            } catch (e: Exception) {
                TransportLogRepository.logFailure(
                    transport = TransportType.FIREBASE,
                    direction = TransportDirection.SEND,
                    messageId = alert.messageId,
                    details = "Firebase send error: ${e.message}"
                )
                Log.w(TAG, "V2N: Firebase send failed: ${e.message}")
            }
        }

        Log.d(TAG, "Dual broadcast: ${alert.vehicleType.displayName} at ${alert.latitude}, ${alert.longitude}")
    }

    private fun determineAlertType(location: Location): AlertType {
        return when {
            location.speed < 1f -> AlertType.STATIONARY
            location.speed < 5f -> AlertType.PASSING
            else -> AlertType.APPROACHING
        }
    }

    private fun createAlertPacket(alert: EmergencyAlert): BitchatPacket {
        return BitchatPacket(
            version = 1u,
            type = MessageType.EMERGENCY_ALERT.value,
            senderID = hexStringToByteArray(myPeerId),
            recipientID = SpecialRecipients.BROADCAST,
            timestamp = alert.timestamp.toULong(),
            payload = alert.toPayload(),
            signature = null,
            ttl = EMERGENCY_TTL
        )
    }

    fun processIncomingAlert(packet: BitchatPacket, fromPeerId: String): EmergencyAlert? {
        val receiveTimestamp = System.currentTimeMillis()

        val alert = EmergencyAlert.fromPayload(packet.payload)
        if (alert == null) {
            Log.w(TAG, "Failed to decode emergency alert from $fromPeerId")
            TransportLogRepository.logFailure(
                transport = TransportType.BLE,
                direction = TransportDirection.RECEIVE,
                messageId = "unknown",
                details = "Decode failed from $fromPeerId"
            )
            return null
        }

        if (!deduplicationService.checkAndMark(alert.messageId)) {
            Log.d(TAG, "Duplicate alert ignored: ${alert.messageId}")
            return null
        }

        val bleLatency = receiveTimestamp - alert.timestamp
        TransportLogRepository.logReceive(
            transport = TransportType.BLE,
            messageId = alert.messageId,
            latencyMs = bleLatency,
            payloadBytes = packet.payload.size,
            details = "from=${fromPeerId.take(8)}"
        )

        val myLocation = _currentLocation.value
        val distance = if (myLocation != null) {
            calculateDistance(
                myLocation.latitude, myLocation.longitude,
                alert.latitude, alert.longitude
            )
        } else {
            Float.MAX_VALUE
        }

        val relativeDirection = if (myLocation != null) {
            calculateRelativeDirection(myLocation, alert)
        } else {
            ""
        }

        val receivedAlert = ReceivedAlert(
            alert = alert,
            distanceMeters = distance,
            relativeDirection = relativeDirection
        )

        addReceivedAlert(receivedAlert)

        delegate?.onEmergencyAlertReceived(receivedAlert)

        Log.i(TAG, "Received emergency alert: ${alert.vehicleType.displayName} at ${distance}m (latency: ${bleLatency}ms)")

        return alert
    }

    private fun addReceivedAlert(alert: ReceivedAlert) {
        val currentList = _receivedAlerts.value.toMutableList()

        currentList.removeAll { !it.isValid }

        currentList.removeAll { it.alert.senderPeerId == alert.alert.senderPeerId }

        currentList.add(alert)

        currentList.sortBy { it.distanceMeters }

        _receivedAlerts.value = currentList
    }

    fun cleanupExpiredAlerts() {
        val currentList = _receivedAlerts.value.toMutableList()
        val sizeBefore = currentList.size
        currentList.removeAll { !it.isValid }

        if (currentList.size != sizeBefore) {
            _receivedAlerts.value = currentList
            Log.d(TAG, "Cleaned up ${sizeBefore - currentList.size} expired alerts")
        }
    }

    private fun calculateDistance(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Float {
        val results = FloatArray(1)
        Location.distanceBetween(lat1, lon1, lat2, lon2, results)
        return results[0]
    }

    private fun calculateRelativeDirection(myLocation: Location, alert: EmergencyAlert): String {
        val bearing = calculateBearing(
            myLocation.latitude, myLocation.longitude,
            alert.latitude, alert.longitude
        )

        val myHeading = myLocation.bearing
        var relativeBearing = bearing - myHeading
        if (relativeBearing < 0) relativeBearing += 360
        if (relativeBearing > 180) relativeBearing -= 360

        return when {
            relativeBearing in -45.0..45.0 -> "ahead"
            relativeBearing in 45.0..135.0 -> "right"
            relativeBearing in -135.0..-45.0 -> "left"
            else -> "behind"
        }
    }

    private fun calculateBearing(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val lat1Rad = Math.toRadians(lat1)
        val lat2Rad = Math.toRadians(lat2)
        val lonDiff = Math.toRadians(lon2 - lon1)

        val x = Math.sin(lonDiff) * Math.cos(lat2Rad)
        val y = Math.cos(lat1Rad) * Math.sin(lat2Rad) -
                Math.sin(lat1Rad) * Math.cos(lat2Rad) * Math.cos(lonDiff)

        var bearing = Math.toDegrees(Math.atan2(x, y))
        if (bearing < 0) bearing += 360

        return bearing
    }

    private fun hexStringToByteArray(hexString: String): ByteArray {
        val result = ByteArray(8) { 0 }
        var tempID = hexString
        var index = 0

        while (tempID.length >= 2 && index < 8) {
            val hexByte = tempID.substring(0, 2)
            val byte = hexByte.toIntOrNull(16)?.toByte()
            if (byte != null) {
                result[index] = byte
            }
            tempID = tempID.substring(2)
            index++
        }

        return result
    }

    fun getDeduplicationService(): AlertDeduplicationService = deduplicationService

    fun shutdown() {
        stopEmergencyBroadcast()

        locationCallback?.let { callback ->
            locationProvider.removeLocationUpdates(callback)
        }
        locationCallback = null

        deduplicationService.shutdown()
        serviceScope.cancel()

        Log.i(TAG, "V2VEmergencyService shutdown")
    }
}

interface V2VEmergencyDelegate {
    fun onEmergencyAlertReceived(alert: ReceivedAlert)
    fun onEmergencyBroadcastStarted(vehicleType: VehicleType)
    fun onEmergencyBroadcastStopped()
}
